lib/host/wasi/inode-linux.cpp

```diff
@@ -751,6 +762,87 @@ WasiExpect<Poller> INode::pollOneoff(__wasi_size_t NSubscriptions) noexcept {
     return WasiUnexpect(__WASI_ERRNO_NOMEM);
   }
 }
+WasiExpect<void>
+INode::getAddrinfo(const char *NodeStr, const char *ServiceStr,
+                   const __wasi_addrinfo_t &Hint, uint32_t MaxResLength,
+                   std::vector<struct __wasi_addrinfo_t *> &WasiAddrinfoArray,
+                   std::vector<struct __wasi_sockaddr_t *> &WasiSockaddrArray,
+                   std::vector<char *> &AiAddrSaDataArray,
+                   std::vector<char *> &AiCanonnameArray,
+                   /*Out*/ __wasi_size_t *ResLength) noexcept {
+  struct addrinfo SysHint;
+  struct addrinfo *SysResPtr = nullptr;
+  struct addrinfo *SysResItem = nullptr;
+  SysHint.ai_flags = Hint.ai_flags;
+  SysHint.ai_family = Hint.ai_family;
+  SysHint.ai_socktype = Hint.ai_socktype;
+  SysHint.ai_protocol = Hint.ai_protocol;
+  SysHint.ai_addrlen = Hint.ai_addrlen;
+  SysHint.ai_addr = NULL;
+  SysHint.ai_canonname = NULL;
+  SysHint.ai_next = NULL;
+
+  if (auto Res = ::getaddrinfo(NodeStr, ServiceStr, &SysHint, &SysResPtr);
+      unlikely(Res < 0)) {
+    return WasiUnexpect(fromErrNo(Res));
+  }
+  // calculate ResLength
+  if (*ResLength = calculateAddrinfoLinkedListSize(SysResPtr);
+      *ResLength > MaxResLength) {
+    *ResLength = MaxResLength;
+  }
+
+  SysResItem = SysResPtr;
+  for (uint32_t Idx = 0; Idx < (*ResLength); Idx++) {
```

Comment:
Use reference instead of using index access to the same element.
`auto &CurAddrinfo = WasiAddrinfoArray[Idx]`;